lakeshore336/system_tests/lewis_emulators/Lksh336/device.py

```diff
@@ -0,0 +1,258 @@
+from collections import OrderedDict
+from .states import DefaultState
+from lewis.devices import StateMachineDevice
+
+
+NUM_OUTPUTS = 4
+INPUTS = ['A', 'B', 'C', 'D']
+
+
+class Ramp(object):
+    """
+    Class representing an output ramp's rate and status.
+    """
+    def __init__(self) -> None:
+        self.rate = 0.0
+        self.status = 0
+
+class PID(object):
+    """
+    Class representing a PID.
+    """
+    def __init__(self):
+        self.p = 0
+        self.i = 0
+        self.d = 0
+
+class OutMode(object):
+    """
+    Class holding the output mode, control input, and powerup.
+    """
+    def __init__(self):
+        self.mode = 0
+        self.control_input = 0
+        self.powerup = 0
+
+class AlarmStatus(object):
+    """
+    Class representing an alarm status.
+    """
+    def __init__(self):
+        self.high = 0
+        self.low = 0
+
+class Alarm(object):
+    """
+    Class representing an alarm.
+    """
+    def __init__(self):
+        self.enabled = 0
+        self.high_setpoint = 0
+        self.low_setpoint = 0
+        self.deadband = 0
+        self.latching = 0
+        self.audible = 0
+        self.visible = 0
+
+class CurveHeader(object):
+    """
+    Class representing the header information of a curve.
+    """
+    def __init__(self):
+        # Needs to be 15 characters.
+        self.name = "".rjust(15, "#")
+        # Needs to be 10 characters.
+        self.serial_number = "".rjust(10, "#")
+        self.data_format = 0
+        self.temperature_limit = 0.0
+        self.temperature_coefficient = 0
+
+class InputType(object):
+    """
+    Class representing the input type parameters.
+    """
+    def __init__(self) -> None:
+        self.sensor_type = 0
+        self.auto_range_setting = 0
+        self.range = 0
+        self.compensation = 0
+        self.units = 0
+
+
+class SimulatedLksh336(StateMachineDevice):
+
+    def _initialize_data(self):
+        self.connected = True
+
+        self.id = ""
+        self.output_heater_values = [0] * NUM_OUTPUTS
+        self.output_analog_outputs = [0] * NUM_OUTPUTS
+        self.output_setpoints = [0.0] * NUM_OUTPUTS
```

Comment:
Sorry, I think I might not have been entirely clear when I explained this to you the other day.

What I was thinking of is something like:

```python
class Output(object):
    def __init__(self):
        self.p = 0
        self.i = 0
        self.d = 0
        self.ramprate = 0
        self.mode = 0
        <all other properties relevant to a single output here>
```

So that in our main class, `_initialize_data` can look like:

```python
class SimulatedLksh336(StateMachineDevice):
    def _initialize_data(self):
        self.connected = True
        outputs = [Output() for _ in range(4)]
        inputs = [Input() for _ in range(4)]
```

Which reduces the duplication of having a bunch of lists, each of which have to implicitly be the same length etc. Basically the goal is to encapsulate all of the variables that each output has together in a single class. Same for inputs.